// @angular/compiler/esm2022/src/template/pipeline/src/phases/resolve_i18n_icu_placeholders.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as i18n from '../../../../i18n/i18n_ast';
import * as ir from '../../ir';
/**
 * Resolves placeholders for element tags inside of an ICU.
 */
export function resolveI18nIcuPlaceholders(job) {
    for (const unit of job.units) {
        for (const op of unit.create) {
            if (op.kind === ir.OpKind.I18nContext && op.contextKind === ir.I18nContextKind.Icu) {
                for (const node of op.message.nodes) {
                    node.visit(new ResolveIcuPlaceholdersVisitor(op.postprocessingParams));
                }
            }
        }
    }
}
/**
 * Visitor for i18n AST that resolves ICU params into the given map.
 */
class ResolveIcuPlaceholdersVisitor extends i18n.RecurseVisitor {
    constructor(params) {
        super();
        this.params = params;
    }
    visitContainerPlaceholder(placeholder) {
        // Add the start and end source span for container placeholders. These need to be recorded for
        // elements inside ICUs. The slots for the nodes were recorded separately under the i18n
        // block's context as part of the `resolveI18nElementPlaceholders` phase.
        if (placeholder.startName && placeholder.startSourceSpan &&
            !this.params.has(placeholder.startName)) {
            this.params.set(placeholder.startName, [{
                    value: placeholder.startSourceSpan?.toString(),
                    subTemplateIndex: null,
                    flags: ir.I18nParamValueFlags.None
                }]);
        }
        if (placeholder.closeName && placeholder.endSourceSpan &&
            !this.params.has(placeholder.closeName)) {
            this.params.set(placeholder.closeName, [{
                    value: placeholder.endSourceSpan?.toString(),
                    subTemplateIndex: null,
                    flags: ir.I18nParamValueFlags.None
                }]);
        }
    }
    visitTagPlaceholder(placeholder) {
        super.visitTagPlaceholder(placeholder);
        this.visitContainerPlaceholder(placeholder);
    }
    visitBlockPlaceholder(placeholder) {
        super.visitBlockPlaceholder(placeholder);
        this.visitContainerPlaceholder(placeholder);
    }
}
